### ISSUE TITLE
[clang] Warning-request: calling a function that converts a pointer argument to bool parameter should issue a warning

### ISSUE THREAD

**Issue by zhihaoy** [llvm/llvm-project#90794]
In an overload set that involves `bool`, it's too easy to fall into the trap where the users are meant to pass the pointer or string literal argument to a more appropriate candidate but end up passing it to the candidate that takes `bool`. A typical case involves
```cpp
void bar(bool);
void bar(std::string);
```
Clang introduced [-Wstring-conversion](https://clang.llvm.org/docs/DiagnosticsReference.html#id877) to diagnose the case where the argument is a string literal,
```cpp
bar("");
```
but not overwise.
```cpp
bar(sz);
```
The non-literal case should also be diagnosed.
The diagnosis should also not be limited to boolean conversion from pointer to strings. Any pointers can fall into this trap:
```cpp
void foo(bool);
void foo(void *);

void test1(T *sp)
{
    foo(sp); // should give a warning
}
```

If we have concerns about the situation where calling a function that takes `bool` with a pointer can appear frequently in practice, we could limit the scope to overload resolution. In other words, if overload resolution calls a candidate that converts a pointer argument to a bool parameter, Clang issues a warning, but this will be more involved.

Repro: https://godbolt.org/z/WbsMnYnW9

**Comment by erichkeane** [comment]
I'm not sure what you mean by the "only in overload resolution", but the action of using the implicit conversion of pointer-to-bool is something code does A LOT and intentionally, and knowingly.  I suspect the false-positive rate here would make this a diagnostic we could never turn on by default.

Even if you only did it in cases where overload resolution contained a candidate that was rejected as a pointer, that would have a significant false-positive rate as well, as the use of that behavior (treating ONE pointer type as a special case, and the rest as bool) is reasonably common as well.

**Comment by zhihaoy** [comment]
> 
> [...] the use of that behavior (treating ONE pointer type as a special case, and the rest as bool) is reasonably common as well.

Sounds like a bug. Do you have an example? Even if it is not a bug, the warnings' purpose is not exactly to reveal bugs as accurately as possible. It's possible that code counting on these has room for improvement and can benefit from the warnings. Considering the actual bugs' harmfulness, some users might want to turn this warning on.

**Comment by erichkeane** [comment]
> > [...] the use of that behavior (treating ONE pointer type as a special case, and the rest as bool) is reasonably common as well.
> 
> Sounds like a bug. Do you have an example? Even if it is not a bug, the warnings' purpose is not exactly to reveal bugs as accurately as possible. It's possible that code counting on these has room for improvement and can benefit from the warnings. Considering the actual bugs' harmfulness, some users might want to turn this warning on.

I've seen the pattern done quite a bit on purpose, particularly for a generic 'visitor' pattern, but also for SFINAE trickery. It isn't a bug, it is an intentional choice made while understanding the semantics of boolean vs pointer conversions.

As far as diagnostics: Yes, warnings purpose IS to reveal bugs as accurately as possible.  Warnings that have high reasonably measurable rates don't raise to the point of being enabled by default (any warning that has ANY amount of false-positives typically doesn't meet our requirements), and warnings we won't turn on by default no longer meet our requirements for inclusion in the compiler (per @AaronBallman ).

Frankly, this sort of warning seems like it would best belong in a clang-tidy sort of warning unless we could make it particularly strictly defined as to not be disruptive.

**Comment by zhihaoy** [comment]
> 
> I've seen the pattern done quite a bit on purpose, particularly for a generic 'visitor' pattern, but also for SFINAE trickery. It isn't a bug, it is an intentional choice made while understanding the semantics of boolean vs pointer conversions.

No function is actually called in the context of SFINAE. Sure, users want to compute at compile-time using language rules, and we don't issue arithmetic warnings in those contexts, either. I'm still curious what visitor pattern must resolve a pointer to `bool` when `void*` is an obvious choice if I imagine.

> 
> As far as diagnostics: Yes, warnings purpose IS to reveal bugs as accurately as possible. Warnings that have high reasonably measurable rates don't raise to the point of being enabled by default (any warning that has ANY amount of false-positives typically doesn't meet our requirements), and warnings we won't turn on by default no longer meet our requirements for inclusion in the compiler (per @AaronBallman ).

[-Wstring-conversion](https://clang.llvm.org/docs/DiagnosticsReference.html#id877) is not enabled by default, not by `-Wall` either, but I appreciate its existence.

> 
> Frankly, this sort of warning seems like it would best belong in a clang-tidy sort of warning unless we could make it particularly strictly defined as to not be disruptive.

I shared the same impression 10 years ago when I saw a case in unit tests until I saw a case in production affecting customers recently. I personally hope anyone can have a chance to try out this warning if it exists. Being said that, clang-tidy might be a way to collect data, putting the deployment effort and the fact that my employer has Coverity on all the projects aside.

**Comment by firewave** [comment]
> Frankly, this sort of warning seems like it would best belong in a clang-tidy sort of warning unless we could make it particularly strictly defined as to not be disruptive

This check already exists in clang-tidy: https://clang.llvm.org/extra/clang-tidy/checks/readability/implicit-bool-conversion.html. Enabling `AllowPointerConditions` you will get the desired warnings.

**Comment by zhihaoy** [comment]
> 
> This check already exists in clang-tidy: https://clang.llvm.org/extra/clang-tidy/checks/readability/implicit-bool-conversion.html. Enabling `AllowPointerConditions` you will get the desired warnings.

In its current state, this is mostly noisy to me, and the proposed fix comes from a different perspective. The bottom line is that I need to warn on the calls when initializing parameters; more precisely, I care about situations where the outcome of overload resolution "forces" a boolean conversion from pointers. When that happens, the syntax we choose to call does not matter; they are all bad. I want to point fingers at the overload set.

**Comment by chrchr-github** [comment]
For the second example, there is no conversion to `bool`: https://godbolt.org/z/cvxvzj5fP

**Comment by zhihaoy** [comment]
> For the second example, there is no conversion to `bool`: https://godbolt.org/z/cvxvzj5fP

Apologize; should be like this https://godbolt.org/z/ebTs1ravG

I'm not saying that clang-tidy's `readability-implicit-bool-conversion` check cannot point it out; my comment was saying 1. it figures out every conversion without considering the context, and in the end, nothing stands out to warranty a useful fix; 2. it proposes a wrong fix because the idea that backs up this check comes from a different perspective.

**Comment by AaronBallman** [comment]
The reason `-Wstring-conversion` behaves the way it does is because converting a string literal to a `bool` is often not what the developer wanted to do because it always converts to the value `true` and there are more clear ways to write that code. However, with non-literal pointers, it's not as clear whether the developer intends the conversion to `bool` or not and so there's a risk of a significantly higher false-positive rate. I would be surprised if the diagnostic behavior was based on an overload set because the language already gives a way to achieve what you'd like in that case: https://godbolt.org/z/8nezh5Kf9

**Comment by zhihaoy** [comment]
> However, with non-literal pointers, it's not as clear whether the developer intends the conversion to `bool` or not and so there's a risk of a significantly higher false-positive rate.

Yes, and I do not consider a context-less warning would give me a practical outcome.

> I would be surprised if the diagnostic behavior was based on an overload set because the language already gives a way to achieve what you'd like in that case: https://godbolt.org/z/8nezh5Kf9

The language feature is exactly what I'm looking for **if** I have a warning with a suggested fix -- it might not be a conventional suggested fix as it would clear the warning but leave an error. Regardless, the customers face a quiet compiler right now because we do not have such a warning in the first place. We've been teaching users that "the language gives you this," and I've seen users who avoid templates in all forms being happy to adopt the template + `= delete` change. But the problem of "you don't know what you don't know" continues to hunt users. A novice does not know there is a trap until being hit once; an experienced user may not be the person who authored the problematic overload set.
